### PR TITLE
Clearer error messages for invalid aesthetics

### LINKED ISSUE #3060: Better error messages for incorrect aesthetic mappings
I encountered some cryptic error messages that pop up in response to fairly normal user mistakes, where the user maps an aesthetic to a data column that doesn't exist. A better error message could help a lot here.

My code example below first shows the desired output and then two incorrect ways in which a user might try to generate it.

``` r
library(ggplot2)

# desired plot
ggplot(mtcars, aes(mpg, 1)) +
  geom_tile(aes(fill = stat(density)), stat = "density") +
  geom_point(position = "jitter")
```

``` r

# wrong, `fill` aesthetic mapped to stat-generated data column that
# doesn't exist in `geom_point()` layer
ggplot(mtcars, aes(mpg, 1, fill = stat(density))) +
  geom_tile(stat = "density") +
  geom_point(position = "jitter")
#> Error in is.finite(x): default method not implemented for type 'closure'

# wrong, `fill` aesthetic mapped to data column that doesn't exist
ggplot(mtcars, aes(mpg, 1, fill = density)) +
  geom_tile(stat = "density") +
  geom_point(position = "jitter")
#> Don't know how to automatically pick scale for object of type function. Defaulting to continuous.
#> Error: Column `fill` must be a 1d atomic vector or a list
```

<sup>Created on 2019-01-07 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

### PR DESCRIPTION
**Pull request by clairemcwhite**
Fixes #3060

Previously, if a stat wasn't valid for layers, the error message was confusing
```{r}
# wrong, `fill` aesthetic mapped to stat-generated data column that
# doesn't exist in `geom_point()` layer
ggplot(mtcars, aes(mpg, 1, fill = stat(density))) +
  geom_tile(stat = "density") +
  geom_point(position = "jitter")
#> Error in is.finite(x): default method not implemented for type 'closure'
```

The new error message for this code is now
```{r}
 #> Error: Aesthetics must be valid computed stats: fill. Did you map your stat in the wrong layer? 
```
In the same vein, if a function like density was used as an aesthetic, the error message is clarified  from the old:

```{r}
# wrong, `fill` aesthetic mapped to data column that doesn't exist
ggplot(mtcars, aes(mpg, 1, fill = density)) +
  geom_tile(stat = "density") +
  geom_point(position = "jitter")
#> Don't know how to automatically pick scale for object of type function. Defaulting to continuous.
#> Error: Column `fill` must be a 1d atomic vector or a list
```

to the new error message below that prompts the user to use stat(density)

```{r}
#> Don't know how to automatically pick scale for object of type function. Defaulting to continuous.
#> Error: Aesthetics must be valid data columns: fill. Did you forget to add stat()? 
```

This was done by adding two checks in R/layer.r 

For the first change:
```{r}
    # Check that all columns in aesthetic stats are valid data
    nondata_stat_cols <- check_nondata_cols(stat_data)
    if(length(nondata_stat_cols) > 0){
      msg <- paste0(
        "Aesthetics must be valid computed stats: ", nondata_stat_cols,
        ". Did you map your stat in the wrong layer?"
      )
      stop(msg, call. = FALSE)
    }
```

and for the second change:

```{r}
    nondata_cols <- check_nondata_cols(evaled) 
    if(length(nondata_cols) > 0){
      msg <- paste0(
        "Aesthetics must be valid data columns: ", nondata_cols,
        ". Did you forget to add stat()?"
      )
      stop(msg, call. = FALSE)
      }
```

Both use a new R/utilities.R function `check_nondata_cols()`
```{r}
# This function checks that all columns of a dataframe `x` are data and
# returns the names of any columns that are not.
# We define "data" as atomic types or lists, not functions or otherwise
check_nondata_cols <- function(x) {
  idx <- (vapply(x, function(x) rlang::is_atomic(x) || rlang::is_list(x), logical(1)))
  names(x)[which(!idx)]
}
```

There are two test in tests/testthat/test-layer.r for these two situations
```{r}

test_that("function aesthetics are wrapped with stat()", {
  df <- data_frame(x = 1:10)
  expect_error(
    ggplot_build(ggplot(df, aes(density)) + geom_tile(stat = "density")),
    "Aesthetics must be valid data columns:"
  )
})

test_that("computed stats are in appropriate layer", {
  df <- data_frame(x = 1:10)
  expect_error(
    ggplot_build(ggplot(df, aes(x = x, stat(density))) + geom_tile(stat = "density") + geom_point()),
    "Aesthetics must be valid computed stats:"
  )
})
```

### CODE REVIEW

**Comment by clauswilke** [comment]
@thomasp85 Could you take a look at the code? It looks good to me but would be good to get a second pair of eyes on it.

@batpigandme Would you mind looking this over for the phrasing of the error messages?

**Comment by thomasp85** [comment]
It appears to be good, but I can for the love of me not shake the feeling that it will error out in some fringe cases where it should not...

But I'm pretty sure that's just me — can't think of any instances where data could be other than atomic or recursive...

**Comment by thomasp85** [comment]
@batpigandme if you are ok with the wording you can approve and merge

File: R/utilities.r [hunk 1]
```diff
@@ -354,6 +354,14 @@ is.discrete <- function(x) {
   is.factor(x) || is.character(x) || is.logical(x)
 }
 
+# This function checks that all columns of a dataframe `x` are data and
+# returns the names of any columns that are not.
+# We define "data" as atomic types or lists, not functions or otherwise
+check_nondata_cols <- function(x) {
+  idx <- (vapply(x, function(x) rlang::is_atomic(x) || rlang::is_list(x), logical(1)))
```
Comment (hadley) [inline R/utilities.r]: This  could use `rlang::is_vector()`

File: R/layer.r [hunk 2]
```diff
@@ -280,6 +289,17 @@ Layer <- ggproto("Layer", NULL,
     env$stat <- stat
 
     stat_data <- new_data_frame(lapply(new, rlang::eval_tidy, data, env))
+
+    # Check that all columns in aesthetic stats are valid data
+    nondata_stat_cols <- check_nondata_cols(stat_data)
+    if(length(nondata_stat_cols) > 0){
```
Comment (hadley) [inline R/layer.r]: Missing space after `if`

File: R/layer.r [hunk 3]
```diff
@@ -229,6 +229,15 @@ Layer <- ggproto("Layer", NULL,
     evaled <- lapply(aesthetics, rlang::eval_tidy, data = data)
     evaled <- compact(evaled)
 
+    nondata_cols <- check_nondata_cols(evaled)
+    if(length(nondata_cols) > 0){
+      msg <- paste0(
+        "Aesthetics must be valid data columns: ", nondata_cols,
+        ". Did you forget to add stat()?"
```
Comment (hadley) [inline R/layer.r]: Is forgetting `stat()` really the most likely cause?

Comment (clauswilke) [reply R/layer.r]: There's two ways to get to this point: Forgetting `stat()` or mistyping a column name in such way that the result is a function, e.g. typing `mean` when the data column is called `Mean`. How about: `"Did you mistype the name of a data column or forget to add stat()?"`

Comment (hadley) [reply R/layer.r]: Can we display the thing that they actually typed? Then we could hint to check that `blah` is a column in the data frame.

**Comment by clauswilke** [comment]
> It appears to be good, but I can for the love of me not shake the feeling that it will error out in some fringe cases where it should not...

I discussed this with @hadley at the developers day and the more I think about it the more I think it's right. Note that we're already performing a similar check on all incoming data, here:

https://github.com/tidyverse/ggplot2/blob/e9d4e5dd599b9f058cbe9230a6517f85f3587567/R/utilities.r#L440-L444

Though now that I think about it, maybe this check and the new check in this PR should be combined into one? Do we need to call tibble to validate the input data or could we simply run something like the `check_nondata_cols()` defined here?

**Comment by thomasp85** [comment]
@clairemcwhite do you want to have a go at finishing this PR off?

**Comment by clairemcwhite** [comment]
I can take a shot this weekend

File: R/utilities.r (hunk 1 above)
Comment (clairemcwhite) [reply R/utilities.r]: Done

```idx <- (vapply(x, function(x) rlang::is_vector(x), logical(1)))```

File: R/layer.r (hunk 2 above)
Comment (clairemcwhite) [reply R/layer.r]: Done

File: R/layer.r (hunk 3 above)
Comment (clairemcwhite) [reply R/layer.r]: I don't know how to get it to print the actual typed statement, but now it prints the problematic aesthetic (along with the tilde). 

Situation 1: 
```ggplot(mtcars, aes(x = mpg, stat(density))) + geom_tile(stat = "density") + geom_point()```

Error: Aesthetics must be valid computed stats: `~stat(density)`. Did you map your stat in the wrong layer? 

Situation 2:
```
ggplot(mtcars, aes(x = mpg, fill = density)) + geom_tile(stat = "density") + geom_point()
```
Don't know how to automatically pick scale for object of type function. Defaulting to continuous.
 Error: Aesthetics must be valid data columns: `~density`. Did you mistype the name of a data column or forget to add stat()?

File: R/layer.r [hunk 4]
```diff
@@ -229,6 +229,15 @@ Layer <- ggproto("Layer", NULL,
     evaled <- lapply(aesthetics, rlang::eval_tidy, data = data)
     evaled <- compact(evaled)
 
+    nondata_cols <- check_nondata_cols(evaled)
+    if (length(nondata_cols) > 0) {
+      msg <- paste0(
+        "Aesthetics must be valid data columns: `", deparse(aesthetics[[nondata_cols]]),
```
Comment (clauswilke) [inline R/layer.r]: I think it would be better to use `as_label()` instead of `deparse()`.

File: R/layer.r [hunk 5]
```diff
@@ -280,6 +289,17 @@ Layer <- ggproto("Layer", NULL,
     env$stat <- stat
 
     stat_data <- new_data_frame(lapply(new, rlang::eval_tidy, data, env))
+
+    # Check that all columns in aesthetic stats are valid data
+    nondata_stat_cols <- check_nondata_cols(stat_data)
+    if (length(nondata_stat_cols) > 0) {
+      msg <- paste0(
+        "Aesthetics must be valid computed stats: `", deparse(aesthetics[[nondata_stat_cols]]),
```
Comment (clauswilke) [inline R/layer.r]: See above.

**Comment by clauswilke** [comment]
As I said in the code comments, I think you should use `as_label()` instead of `deparse()`. However, be aware that there's a dependency with #3242, which isn't merged yet. For now, you should write `rlang::as_label()`. We can then wait merging this PR until #3242 is merged, and then change `rlang::as_label()` to `as_label()` here and merge this.

**Comment by clairemcwhite** [comment]
Deparse is now replaced with rlang::as_label. Current behavior is : 

```ggplot(mtcars, aes(x = mpg, stat(density))) + geom_tile(stat = "density") + geom_point()```

Error: Aesthetics must be valid computed stats: `stat(density)`. Did you map your stat in the wrong layer? 

```
ggplot(mtcars, aes(x = mpg, fill = density)) + geom_tile(stat = "density") + geom_point()
```
Don't know how to automatically pick scale for object of type function. Defaulting to continuous.
 Error: Aesthetics must be valid data columns: `density`. Did you mistype the name of a data column or forget to add stat()?

File: NEWS.md [hunk 6]
```diff
@@ -8,6 +8,8 @@
 
 * `coord_map()` now can have axes on the top and right (@karawoo, #3042).
 
+* Clearer error messages for innappropriate aesthetics. (@clairemcwhite, #3060).
```
Comment (clauswilke) [inline NEWS.md]: There shouldn't be a period after "aesthetics" (only at the very end).

File: R/layer.r [hunk 7]
```diff
@@ -229,6 +229,15 @@ Layer <- ggproto("Layer", NULL,
...
     evaled <- compact(evaled)
 
+    nondata_cols <- check_nondata_cols(evaled)
+    if (length(nondata_cols) > 0) {
+      msg <- paste0(
+        "Aesthetics must be valid data columns: `", rlang::as_label(aesthetics[[nondata_cols]]),
+        "`. Did you mistype the name of a data column or forget to add stat()?"
+      )
+      stop(msg, call. = FALSE)
+      }
```
Comment (clauswilke) [inline R/layer.r]: The indentation of the closing } seems wrong (two spaces too much).

File: R/layer.r [hunk 8]
```diff
@@ -229,6 +229,15 @@ Layer <- ggproto("Layer", NULL,
     evaled <- lapply(aesthetics, rlang::eval_tidy, data = data)
     evaled <- compact(evaled)
 
+    nondata_cols <- check_nondata_cols(evaled)
+    if (length(nondata_cols) > 0) {
+      msg <- paste0(
+        "Aesthetics must be valid data columns: `", rlang::as_label(aesthetics[[nondata_cols]]),
+        "`. Did you mistype the name of a data column or forget to add stat()?"
```
Comment (clauswilke) [inline R/layer.r]: I would remove the backticks around the aesthetics, because it'll look strange if you have more than one.

File: R/layer.r [hunk 9]
```diff
@@ -280,6 +289,17 @@ Layer <- ggproto("Layer", NULL,
     env$stat <- stat
 
     stat_data <- new_data_frame(lapply(new, rlang::eval_tidy, data, env))
+
+    # Check that all columns in aesthetic stats are valid data
+    nondata_stat_cols <- check_nondata_cols(stat_data)
+    if (length(nondata_stat_cols) > 0) {
+      msg <- paste0(
+        "Aesthetics must be valid computed stats: `", rlang::as_label(aesthetics[[nondata_stat_cols]]),
```
Comment (clauswilke) [inline R/layer.r]: As above, remove backticks around aesthetics.

File: R/layer.r (hunk 7 above)
Comment (clairemcwhite) [reply R/layer.r]: Fixed

File: R/layer.r (hunk 9 above)
Comment (clairemcwhite) [reply R/layer.r]: Backticks removed

File: R/layer.r (hunk 8 above)
Comment (clairemcwhite) [reply R/layer.r]: Backticks removed

File: NEWS.md (hunk 6 above)
Comment (clairemcwhite) [reply NEWS.md]: Fixed. Also fixed wrong spelling of inappropriate.

**Comment by clauswilke** [comment]
Claire, could you check what happens when there are multiple problematic aesthetics in one layer? I just realized that that case may not work yet. It probably needs a `paste0()` and a `vapply()`: 

    paste0(vapply(aesthetics[[nondata_stat_cols]], rlang::as_label, character(1)), collapse = ", ")

Also make sure the test functions check against the final text in the error message. You can probably just delete the final colon.

**Comment by yutannihilation** [comment]
#3242 is merged now.

File: NEWS.md [hunk 10]
```diff
@@ -81,6 +81,10 @@ core developer team.
 
 * ggplot2 now works in Turkish locale (@yutannihilation, #3011).
 
+* Clearer error messages for inappropriate aesthetics (@clairemcwhite, #3060).
+
+* `geom_rug()` gains an "outside" option to allow for moving the rug tassels to outside the plot area. (@njtierney, #3085)
```
Comment (clauswilke) [inline NEWS.md]: Do we have an idea of where this news item comes from? It should have been present in NEWS.md already, but it isn't, so I'm confused about both why it was lost and how it made its reappearance here.

Comment (clauswilke) [reply NEWS.md]: Ah, it was moved in this commit: https://github.com/tidyverse/ggplot2/commit/230e8f78ae54d255aa02ad2f2d7efca9c46905d3 so you can just delete it here.

Comment (clairemcwhite) [reply NEWS.md]: Got it, will do in next commit.

**Comment by clauswilke** [comment]
Claire, just a reminder, the deadline to merge this for the 3.2.0 release is April 30. It would be great if you could complete this by this deadline. Thanks!

**Comment by clairemcwhite** [comment]
@clauswilke 

Multiple problem aesthetics now work. 

Case 1:
> ggplot(mtcars, aes(x = mpg, fill = density, color = density)) + geom_tile(stat = "density") + geom_point(stat = "density")
Don't know how to automatically pick scale for object of type function. Defaulting to continuous.
Don't know how to automatically pick scale for object of type function. Defaulting to continuous.
 Error: Aesthetics must be valid data columns. Problematic aesthetic(s): density, density. 
Did you mistype the name of a data column or forget to add stat()? 

Case 2:

> ggplot(mtcars, aes(x = mpg, fill = stat(density), colour = stat(density))) + geom_tile(stat = "density") + geom_point()
 Error: Aesthetics must be valid computed stats. Problematic aesthetic(s): stat(density), stat(density). 
Did you map your stat in the wrong layer? 


The only improvement I can think of would be to have the error include the type of aesthetic, like: "Problematic aesthetics(s): fill = stat(density), color = stat(density)". 

The variable "nondata_cols" is a vector c("fill", "color"), so there should be a quick way to construct this kind of error statement.

**Comment by clauswilke** [comment]
Very good! Final set of comments, I promise. :-)

1. To improve the error message, I think you can replace the `function(x) {...}` statement with:

    function(x) {paste0(x, " = ", as_label(aesthetics[[x]]))}

2. I would use the case with multiple broken aesthetics in the unit tests, and check that the aesthetics are listed correctly (after addressing my point 1).

3. Do you need two geoms to create the multi-aesthetics case, or is one geom sufficient as long as you have `aes(color = ..., fill = ...)`?

4. What happens when both types of mistakes appear in the same plot?

**Comment by clairemcwhite** [comment]
1. Cool, the new function is function(x) {paste0(x, " = ", as_label(aesthetics[[x]]))}

Case 1:
```
ggplot(mtcars, aes(x = mpg, fill = density, color = density)) + geom_tile(stat = "density") + geom_point(stat = "density")
Don't know how to automatically pick scale for object of type function. Defaulting to continuous.
Don't know how to automatically pick scale for object of type function. Defaulting to continuous.
 Error: Aesthetics must be valid data columns. Problematic aesthetic(s): fill = density, colour = density. 
Did you mistype the name of a data column or forget to add stat()? 
```

Case 2:
```
ggplot(mtcars, aes(x = mpg, fill = stat(density), color = stat(density))) + geom_tile(stat = "density") + geom_point()
 Error: Aesthetics must be valid computed stats. Problematic aesthetic(s): fill = stat(density), colour = stat(density). 
Did you map your stat in the wrong layer? 
```
One note is that this approach anglicizes my "color" to "colour"

2. I changed the tests to be more like what you suggested.

3. One geom is sufficient "ggplot(mtcars, aes(fill = stat(density), color = stat(density))) + geom_point()" and "ggplot(mtcars, aes(fill = density, color = density)) + geom_point()" produce the two errors

4. You get the errors in the order that they appear in layer.R. Remove the first error, and you get the second error
```
ggplot(mtcars, aes(x = mpg, fill = stat(density), color = density)) + geom_point()
Don't know how to automatically pick scale for object of type function. Defaulting to continuous.
 Error: Aesthetics must be valid data columns. Problematic aesthetic(s): colour = density. 
Did you mistype the name of a data column or forget to add stat()? 
# get rid of color = density
> ggplot(mtcars, aes(x = mpg, fill = stat(density))) + geom_point()
 Error: Aesthetics must be valid computed stats. Problematic aesthetic(s): fill = stat(density). 
Did you map your stat in the wrong layer? 
```

**Comment by clauswilke** [comment]
Super, thanks! Assuming the automated regression tests pass I'll merge.

**Comment by clauswilke** [comment]
The AppVeyor build fails, for reasons that I can't discern. It doesn't seem to have anything to do with the changes that were made in this PR. The tests that break are visual tests for the date scale.

@thomasp85, any idea what's happening? Can we merge and hope for the best?

**Comment by thomasp85** [comment]
It has been going on since 5aacacb... I think it is fair to assume it has nothing to do with this PR

**Comment by clauswilke** [comment]
Thanks!

**Pull request merged**